### PR TITLE
[PP-150] Done

### PR DESCRIPTION
**Pull request by Claudia-Partnersense**
**Inledande ord:** Måste få säga att jag är riktigt stolt över hur detta blev. Om jag får säga det själv den bästa koden jag (Claude ;)) skrivit, tester på alla logic metoder (dvs inga integrationtester), docs på alla klasser och metoder, tydlig separation på klasser med endast relevanta metoder i klasserna. 

 Tog längre tid än planerat men skyller det på ringelrostigheten efter varit borta ett tag samt invänning med att börja arbeta så tight med Claude. 

Jag har sammanställt allt nedan, jag skulle rekomendera att kolla igenom texten nedan och sedan dra ner repot och köra det lokalt enligt instruktionerna i **"Running Locally"**, det ska gå hur smidigt som helst och gör det inte det återkoppla gärna för vill att det ska vara enkelt att köra denna koden utan strul eller setup mer än kopiera från bitwarden. 

# DataFeedWatch Feed Builder Service

## Overview
This PR introduces a new generalized feed builder service that consolidates product data from Norce into a unified multi-market feed for DataFeedWatch consumption. The service is designed to be configurable and support multiple clients, cultures, and sales areas through configuration in azure rather than hardcoded values.

### Key Features
- Multi-market support through configurable culture and sales area filters
- Separation of price-specific and culture-specific product attributes
- Unified feed generation combining data from multiple markets
- Azure Blob Storage integration for feed storage
- Comprehensive error handling and logging
- Extensive test coverage (87 tests in FeedService, 75 tests in SharedLib)

## Architecture
The service is structured into distinct layers with clear separation of concerns:

### Core Services
- `CultureFeedGenerationService`: Handles culture-specific data (names, descriptions)
- `PriceFeedGenerationService`: Manages price and sales area specific data
- `SalesAreaConfigurationService`: Configures and validates sales area configurations
- `CultureConfigurationService`: Configures and validates  culture-specific configurations
- `AzureStorageService`: Handles Azure Blob Storage operations

### Helpers and Extensions
- Various extension classes for mapping and validation logic: I have seperated the logic and the fetching and sending of data meticulously in this project, therefore the extension classes. These classes should only contain pure logic, this to make testing easier and more metainible, and also to easier maintain but also monitor the code, if the error is in a service class its a integration error, if it is in an extension class its a logic error. I hope this is a good way to do it.
- `ProductFeedUnifierHelper`: Combines data from different markets into unified feed

### Data Models
- Clear separation between internal models and DTOs
- Culture-specific models (`CultureConfiguration`)
- Sales area models (`SalesAreaConfiguration`)
- Price-specific models (`PriceProduct`)

## Configuration
All configuration is externalized to Azure:

### Azure Resources
1. **Configuration**: `Callisto-development-app-configuration`
2. **Key Vault**: `callisto-dev-key-vault`
3. **Blob Storage**: Check `callistodevdatafeedwatch - DataStorage - Containers` for example output

### Key Configuration Parameters
- `Options:Base:IncludedCultures`: Filter for supported cultures
- `Options:Base:IncludedSalesAreaIds`: Filter for included sales areas
- Various Norce-specific configurations
- Storage and logging configurations
see azure for more detail configurations 

## Data Flow
1. Get an filter configurations for Cultures and SalesAreas(ink pricelists)
2. Service fetches product price data from Norce trough api, configured sales areas (prices, stock levels)
3. Service fetches product culture data from Norce trough channel feed configured cultures  (names, descriptions)
6. Unified feed is generated combining all market data
7. Feed is uploaded to Azure Blob Storage

### Prerequisites
1. Access to Azure resources
2. .NET 8 SDK
3. Access to Bitwarden for secrets

### Configuration Steps
1. Get necessary secrets from Bitwarden:
   - Check "Callisto launchSettings", Use "LocalDevWithAzureSettings" profile
   - For appsettings see Callisto Backend Application - appsettings.Development

### Running Locally
1. Set environment profile to "LocalDevWithAzureSettings"
2. Service will use Azure App Configuration
3. Run the application using `dotnet run`

## Testing
- Comprehensive test suite using Moq framework
- 87 tests in FeedService project
- 75 tests in SharedLib project
- Coverage includes happy path and error scenarios

## Future Considerations
- Monitor performance with large datasets

## Review Notes
Please pay special attention to:
- Configuration handling and secrets management
- Error handling and logging implementation
- Market-specific data transformation logic
- Test coverage for edge cases

### CODE REVIEW

File: Services/FeedService/FeedService/Domain/Internal/MethodOptions/StorageUploadOptions.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        /// <summary>
+        /// Optional format for the file being uploaded (e.g., "json", "xml", etc.)
+        /// </summary>
+        public string? FileFormat { get; init; }
+
+        /// <summary>
+        /// Whether the file should be publicly accessible
+        /// </summary>
+        public bool IsPublic { get; set; }
```
Comment (Partnersenseadan) [inline Services/FeedService/FeedService/Domain/Internal/MethodOptions/StorageUploadOptions.cs:21]: property should potentially default to false for safer defaults:

Comment (Partnersenseadan) [reply Services/FeedService/FeedService/Domain/Internal/MethodOptions/StorageUploadOptions.cs:21]: public bool IsPublic { get; set; } = false;

File: Services/FeedService/FeedService/Services/StorageServices/AzureStorageService.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                    traceId,
+                    nameof(AzureStorageService),
+                    nameof(LoggingTypes.CheckpointLog),
+                    nameof(UploadAsync),
+                    "Starting file upload to Azure Storage",
+                    nameof(MethodActionLogTypes.Starting)
+                );
+
+                // Validate inputs
+                if (content == null) throw new ArgumentNullException(nameof(content));
```
Comment (Partnersenseadan) [inline Services/FeedService/FeedService/Services/StorageServices/AzureStorageService.cs:45]: good practice seen with proper validation before blob operations, but consider adding content validation before upload:
if (content.Count == 0)
{
    logger.LogWarning("Attempting to upload empty content");
    return false;
}

**Review by Partnersenseadan — CHANGES_REQUESTED** [review]
Well done!!!

Testing 🧪

Excellent test coverage for the new components, particularly ProductFeedUnifierHelperTests.cs.

Positive Aspects ✅

Comprehensive XML documentation on public APIs
Consistent use of cancellation tokens for async operations
Good use of dependency injection
Strong unit test coverage
Clear separation of concerns between services and their extensions

Minor Issues 🔍

Some redundant using statements in various files
Inconsistent line endings in some files
A few magic numbers that could be constants

Recommendations 💡

Consider adding:

XML documentation for extension methods
Interface for ProductFeedUnifierHelper
More logging for successful operations
Configuration validation at startup

File: Services/FeedService/FeedService/Domain/Internal/MethodOptions/StorageUploadOptions.cs (hunk 1 above)
Comment (Claudia-Partnersense) [reply Services/FeedService/FeedService/Domain/Internal/MethodOptions/StorageUploadOptions.cs:21]: fixat

File: Services/FeedService/FeedService/Services/StorageServices/AzureStorageService.cs (hunk 2 above)
Comment (Claudia-Partnersense) [reply Services/FeedService/FeedService/Services/StorageServices/AzureStorageService.cs:45]: fixat

**Review by Partnersenseadan — APPROVED** [review]
